README
README for Sequencer24 0.9.4.3
Chris Ahlstrom
2015-07-24 to 2015-09-16

This package is currently oriented to developers.  Please see the
INSTALL file for a brief description on installation.

The sequencer24 project is a gradual refactoring of seq24 0.9.2
to bring it up to date, fix bugs, consolidate patches, convert
to GTKmm 3, and much more.

The refactoring is done as a series of Git branches (x means it is done, y
means in progress):

x  Legacy:     The original Seq24 0.9.2 with some notes.
x  Debian:     Incorporating Debians packaging support into the project.
x  Catchup:    Incorporating a few patches found around the Internet.
x  Packaging:  Getting the build system up-to-date; bootstrapping; packing.
x  Gtkmm-3:    Porting from gtkmm-2.4 to gtkmm-3.0.  Turned out to be trivial.
x  Document:   Adding thorough in-code documentation and upgraded formatting.
x  Fixes:      More documentation, fixes, and some new features.  Moving
               some fixes to the upcoming Library branch.
   Library:    Refactor the code to separate and unit-test functionality.

The Fixes branch is now "finished". Light documentation, code prettifying,
warning elimination, valgrind clean.  Trying to get a handle on all the
global variables and how things work.  Some fixes will be offloaded to the
Library branch.  In the meantime, enjoy the master branch!

See the ROADMAP file for more information and a kind of TODO list.

Note that the original seq24 is pretty well documented by the following
project, which provides a 60-page PDF file and its LaTeX source code:

   https://github.com/ahlstromcj/seq24-doc.git

However, we have found some documentation mistakes while researching, and
will have to correct those mistakes soon.

The following section is for historicial purposes only.

--------------------------------------
The original seq24 v0.9.2 README
--------------------------------------

How to install? 

    read INSTALL.

How to use seq24?
    
    Documentation on using seq24 is in the SEQ24 file.

Information on modifying, copying?

    This software is free, and released under the GPL.
    This is available in the COPYING file.

Who wrote this?

    Read AUTHORS.

Why did they write this?

    Again, Read SEQ24

What to do with a fresh repository checkout?

    Apply "autoreconf" to get a configure script, then read INSTALL.

# vim: sw=4 ts=4 wm=8 et ft=sh


ChangeLog
2015-09-16  Chris Ahlstrom <ahlstromcj@gmail.com>

	* src/perform.cpp, src/optionsfile.cpp: Fixed a long-standing bug
	in handling keyboard-control changes, setting version to 0.9.4.3.

2015-09-05  Chris Ahlstrom <ahlstromcj@gmail.com>

	* src/perfroll_input.cpp: Added support for Mod4 mode to the Song Editor,
	so that it works in that editor as well as the Pattern Editor, setting
	version to 0.9.4.2.

2015-09-01  Chris Ahlstrom <ahlstromcj@gmail.com>

	* src/seq24.cpp: Added support for configuration files in
	/home/user/.config/sequencer24 when in non-legacy mode, setting
	version to 0.9.4

2015-07-24  Chris Ahlstrom <ahlstromcj@gmail.com>

	* *.*: Porting to new build system, tons of documentation, code cleanup,
	basic fixes, minor new features, for new version 0.9.3 through
	2015-08-29, name change to Sequencer24.

2012-02-24  Guido Scholz  <guido.scholz@bayernline.de>

	* src/perform.*: Register empty process callback for JACK, patch
	provided by Lieven Moors

2010-11-25  Guido Scholz  <guido.scholz@bayernline.de>

	* configure.in: remove unused variables, harmonize option
	enabling/disabling

2010-11-24  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp, src/mainwnd.h: use initialization list for
	members, close pipe on exit, use name label for sreen set name

2010-11-23  Guido Scholz  <guido.scholz@bayernline.de>

	* src/seq24.cpp: read only existing configuration files

2010-11-23  Guido Scholz  <guido.scholz@bayernline.de>

	* src/configfile.cpp, src/configfile.h, src/controllers.h,
	src/event.cpp, src/event.h, src/font.cpp, src/font.h, src/globals.h,
	src/maintime.cpp, src/maintime.h, src/mainwid.cpp, src/mainwid.h,
	src/mainwnd.h, src/midibus.cpp, src/midibus.h,
	src/midibus_portmidi.cpp, src/midibus_portmidi.h, src/midifile.cpp,
	src/mutex.cpp, src/mutex.h, src/options.cpp, src/options.h,
	src/optionsfile.cpp, src/optionsfile.h, src/perfedit.cpp,
	src/perfedit.h, src/perfnames.cpp, src/perfnames.h, src/perform.cpp,
	src/perform.h, src/perfroll.cpp, src/perfroll.h, src/seqdata.cpp,
	src/seqdata.h, src/seqedit.cpp, src/seqedit.h, src/seqevent.cpp,
	src/seqevent.h, src/seqkeys.cpp, src/seqkeys.h, src/seqmenu.cpp,
	src/seqmenu.h, src/seqroll.cpp, src/seqroll.h, src/seqtime.cpp,
	src/seqtime.h, src/sequence.cpp, src/sequence.h, src/userfile.cpp,
	src/userfile.h: remove trailing whitespaces

2010-11-23  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp, src/seq24.cpp: add missing line breaks

2010-11-23  Guido Scholz  <guido.scholz@bayernline.de>

	* man/seq24.1, src/mainwnd.h, src/seq24.cpp: remove -f commandline
	option to be replaced by a simple <filename> argument

2010-11-23  Guido Scholz  <guido.scholz@bayernline.de>

	* src/lash.cpp, src/lash.h, src/seq24.cpp: fix lash connect to
	start after command line parsing

2010-11-22  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp, src/mainwnd.h: add interrupt handlers for
	SIGINT and SIGUSR1 to enable LADISH level 1 support

2010-11-22  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp: fix typo again

2010-11-22  Guido Scholz  <guido.scholz@bayernline.de>

	* src/keybindentry.cpp, src/keybindentry.h, src/Makefile.am,
	src/options.cpp: move keybindentry class to separate file

2010-11-21  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp: fix typo

2010-11-21  Guido Scholz  <guido.scholz@bayernline.de>

	* src/seq24.cpp: change short option for version to V, fix
	compiler warning

2010-11-20  Guido Scholz  <guido.scholz@bayernline.de>

	* man/seq24.1: add missing command line parameters to man page

2010-11-20  Guido Scholz  <guido.scholz@bayernline.de>

	* src/seq24.cpp: add missing command line parameters to help
	message, show also short options, add version option

2010-11-20  Guido Scholz  <guido.scholz@bayernline.de>

	* src/lash.cpp, src/seq24.cpp: first parse command line before
	anything else is done

2010-11-20  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp: fix doubled key event for screen set name line
	(notes edit line)

2010-11-20  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp: rearrange bottom line item layout to get rid of
	whitespace label spacing, add mnemonics for widgets

2010-11-20  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp: rearrange bottom line item addition

2010-11-14  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.cpp: rename box containers to make clear what they
	are responsible for

2010-11-11  Guido Scholz  <guido.scholz@bayernline.de>

	* src/mainwnd.h: remove doubled options header include

2010-11-11  Guido Scholz  <guido.scholz@bayernline.de>

	* src/options.cpp: remove doubled "show_all()" call

2010-11-11  Guido Scholz  <guido.scholz@bayernline.de>

	* INSTALL, configure.in, src/globals.h, src/perform.cpp,
	src/perform.h, src/seq24.cpp: add support for jack session, patch
	provided by Torben Hohn (slightly changed)

2010-11-10  Guido Scholz  <guido.scholz@bayernline.de>

	* src/pixmaps/Makefile.am: add missing file

2010-11-10  Guido Scholz  <guido.scholz@bayernline.de>

	* src/pixmaps, configure.in, src/Makefile.am, src/font.cpp,
	src/mainwnd.cpp, src/perfedit.cpp, src/seqedit.cpp, src/bus.xpm,
	src/pixmaps/bus.xpm, src/collapse.xpm, src/pixmaps/collapse.xpm,
	src/copy.xpm, src/pixmaps/copy.xpm, src/down.xpm,
	src/pixmaps/down.xpm, src/expand.xpm, src/pixmaps/expand.xpm,
	src/font_b.xpm, src/pixmaps/font_b.xpm, src/font_w.xpm,
	src/pixmaps/font_w.xpm, src/key.xpm, src/pixmaps/key.xpm,
	src/learn.xpm, src/pixmaps/learn.xpm, src/learn2.xpm,
	src/pixmaps/learn2.xpm, src/length.xpm, src/pixmaps/length.xpm,
	src/loop.xpm, src/pixmaps/loop.xpm, src/menu_empty.xpm,
	src/pixmaps/menu_empty.xpm, src/menu_full.xpm,
	src/pixmaps/menu_full.xpm, src/midi.xpm, src/pixmaps/midi.xpm,
	src/note_length.xpm, src/pixmaps/note_length.xpm, src/perfedit.xpm,
	src/pixmaps/perfedit.xpm, src/play.xpm, src/pixmaps/play.xpm,
	src/play2.xpm, src/pixmaps/play2.xpm, src/q_rec.xpm,

src/font.cpp
/*
 *  This file is part of seq24/sequencer24.
 *
 *  seq24 is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  seq24 is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with seq24; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/**
 * \file          font.cpp
 *
 *  This module declares/defines the base class for font handling.
 *
 * \library       sequencer24 application
 * \author        Seq24 team; modifications by Chris Ahlstrom
 * \date          2015-07-24
 * \updates       2015-08-23
 * \license       GNU GPLv2 or above
 *
 */

#include "easy_macros.h"
#include "font.h"
#include "pixmaps/font_w.xpm"           /* white on black (inverse video)   */
#include "pixmaps/font_b.xpm"           /* black on white                   */
#include "pixmaps/font_yb.xpm"          /* yellow on black (inverse video)  */
#include "pixmaps/font_y.xpm"           /* black on yellow                  */

/**
 *    Rote default constructor.
 */

font::font ()
 :
    m_pixmap        (nullptr),
    m_black_pixmap  (),
    m_white_pixmap  (),
    m_b_on_y_pixmap (),
    m_y_on_b_pixmap (),
    m_clip_mask     ()
{
   // empty body
}

/**
 *  Initialization function for a window on which fonts will be drawn.
 *  This function loads two pixmaps that contain the characters to be used
 *  to draw text strings.  Both pixmaps provide a 16 x 16 grid of boxes,
 *  and each box contains one of the 256 characters in this font set.
 *
 *  One pixmap has white characters on a black background, and other other
 *  has black characters on a white background.  See the descriptions of
 *  the c_text_x and c_text_y variables in the globals module.
 */

void
font::init (Glib::RefPtr<Gdk::Window> a_window)
{
    m_black_pixmap = Gdk::Pixmap::create_from_xpm
    (
        a_window->get_colormap(), m_clip_mask, font_b_xpm
    );
    m_white_pixmap = Gdk::Pixmap::create_from_xpm
    (
        a_window->get_colormap(), m_clip_mask, font_w_xpm
    );
    m_b_on_y_pixmap = Gdk::Pixmap::create_from_xpm
    (
        a_window->get_colormap(), m_clip_mask, font_y_xpm
    );
    m_y_on_b_pixmap = Gdk::Pixmap::create_from_xpm
    (
        a_window->get_colormap(), m_clip_mask, font_yb_xpm
    );
}

/**
 *  Draws a text string. This function grabs the proper font bitmap,
 *  extracts the current character pixmap from it, and slaps it down where
 *  it needs to be to render the character in the string.
 *
 * \param a_gc
 *      Provides the graphics context for drawing the text using GTK+.
 *
 * \param x
 *      The horizontal location of the text.
 *
 * \param y
 *      The vertical location of the text.
 *
 * \param a_draw
 *      The drawable object on which to draw the text.
 *
 * \param str
 *      The string to draw.  Should use a constant string reference
 *      instead.
 *
 * \param col
 *      The font color to use to draw the string.  The only support values
 *      are font::BLACK and font::WHITE, and the correct colors are
 *      provided by selecting one of two font pixmaps, as described in the
 *      init() function.
 */

void
font::render_string_on_drawable
(
    Glib::RefPtr<Gdk::GC> a_gc,
    int x, int y,
    Glib::RefPtr<Gdk::Drawable> a_draw,
    const char * str,
    font::Color col
)
{
    int length = 0;
    if (not_nullptr(str))
        length = strlen(str);

    /*
     * The width is identical to c_text_x, but the height is not identical
     * to c_text_y.  Using the latter values causes artifacts on the
     * pattern grid.
     */

    int font_w = 6;                     // c_text_x == 6
    int font_h = 10;                    // c_text_y == 12
    if (col == font::BLACK)
        m_pixmap = &m_black_pixmap;
    else if (col == font::WHITE)
        m_pixmap = &m_white_pixmap;
    else if (col == font::BLACK_ON_YELLOW)
        m_pixmap = &m_b_on_y_pixmap;
    else if (col == font::YELLOW_ON_BLACK)
        m_pixmap = &m_y_on_b_pixmap;
    else
        m_pixmap = &m_black_pixmap; // user lied, provide a legal pointer

    for (int i = 0; i < length; ++i)
    {
        unsigned char c = (unsigned char) str[i];
        int pixbuf_index_x = c % 16;    // number of grids horizontally
        int pixbuf_index_y = c / 16;    // number of grids vertically
        pixbuf_index_x *= 9;            // width of grid (letter is 6 pixels)
        pixbuf_index_x += 2;            // add 2 for border?
        pixbuf_index_y *= 13;           // height of grid (letter is 12 pixels)
        pixbuf_index_y += 2;            // add 2 for border?
        a_draw->draw_drawable
        (
            a_gc, *m_pixmap, pixbuf_index_x, pixbuf_index_y,
            x + (i * font_w), y, font_w, font_h
        );
    }
}

/*
 * font.cpp
 *
 * vim: sw=4 ts=4 wm=8 et ft=cpp
 */
